// listlyTest/index.js

var listly = require('../listly');
var mongoskin = require('mongoskin');

var ServerResponseStub = module.exports.ServerResponseStub = function ()
{
	this.setHeader = function (name,value)
	{

	};

	this.send = function(data)
	{

	};
};

var ServerRequestStub = module.exports.ServerRequestStub = function ()
{
	this.params = {};
	this.query = {};
};

var ExpressAppStub = module.exports.ExpressAppStub = function ()
{
	this.get = function (name, handler) { };
	this.post = function (name, handler) { };
	this.put = function (name, handler) { };
};

var ListlyDataServerStub = module.exports.ListlyDataServerStub = function ()
{
	this.app = new ExpressAppStub();
	this.db = new ListlyMongoDBStub();

	this.submissionDAO = new listly.SubmissionDAO(this.db);
	this.listDAO = new listly.ListDAO(this.db);
	this.userDAO = new listly.UserDAO(this.db);

	this.submissionREST = new listly.SubmissionREST(this);
	this.listREST = new listly.ListREST(this);
	this.userREST = new listly.UserREST(this);
};

var ListlyMongoDBStub = module.exports.ListlyMongoDBStub = function ()
{
	this.collection = function (name)
	{
		return this;
	};

	this.find = function ()
	{
		return this;
	};

	this.insert = function ()
	{
		return this;
	};

	this.updateById = function ()
	{
		return this;
	};

	this.toArray = function ()
	{
		return this;
	};
}
