// listly/lib/dao/submissionDAO.js
var mongo = require("mongodb");

var SubmissionDAO = module.exports.SubmissionDAO = function (db)
{
	this.db = db;
};

SubmissionDAO.prototype.findAllSubmissions = function (callback)
{
	this.db.collection("v1_submission").find().toArray(function (err, items)
	{
		callback(items);
	});
};

SubmissionDAO.prototype.findSubmissionsByTitle = function (title, callback)
{
	this.db.collection("v1_submission").find({title:title}).toArray(function (err, items)
	{
		callback(items);
	});
};

SubmissionDAO.prototype.findSubmissionById = function (id, callback)
{
	var submissionId = new mongo.ObjectID(id);

	this.db.collection("v1_submission").find({_id:submissionId}, {limit:1}).toArray(function (err, items)
	{
		callback(items.length > 0 ? items[0] : null);
	});
};

SubmissionDAO.prototype.findSubmissionListsById = function (id, callback)
{
	var submissionId = new mongo.ObjectID(id);

	this.db.collection("v1_submission_to_lists").find({_id:submissionId}, {limit:1}).toArray(function (err, items)
	{
		callback(items.length > 0 ? items[0].value.lists : null);
	});
};

SubmissionDAO.prototype.createSubmission = function (submission, callback)
{
	this.db.collection("v1_submission").insert(submission, function (err)
	{
		callback(err);
	});
};

SubmissionDAO.prototype.updateSubmissionById = function (id, submission, callback)
{
	var submissionId = new mongo.ObjectID(id);

	this.db.collection("v1_submission").updateById(submissionId, submission, function (err)
	{
		callback(err);
	});
};
