// listly/lib/dao/submissionDAO.js

var mongo = require("mongodb");

var SubmissionDAO = module.exports.SubmissionDAO = function (db)
{
	var self = this;

	this.db = db;

	this.findAllSubmissions = function (callback)
	{
		self.db.collection("v1_submission").find().toArray(function (err, items)
		{
			callback(items);
		});
	};

	this.findSubmissionById = function (id, callback)
	{
		var submissionId = new mongo.ObjectID(id);

		self.db.collection("v1_submission").find({_id: submissionId}, {limit:1}).toArray(function (err, items)
		{
			callback(items[0]);
		});
	};

	this.findSubmissionListsById = function (id, callback)
	{
		var submissionId = new mongo.ObjectID(id);

		self.db.collection("v1_submission_to_lists").find({_id: submissionId}, {limit:1}).toArray(function (err, items)
		{
			if (items.length != 1)
			{
				callback(null);
			}
			else
			{
				callback(items[0].value.lists);
			}
		});
	};

	this.findSubmissionsByTitle = function (title, callback)
	{
		self.db.collection("v1_submission").find({title:title}).toArray(function (err, items)
		{
			callback(items);
		});
	};

	this.createSubmission = function (submission, callback)
	{
		self.db.collection("v1_submission").insert(submission, function (err, submissions)
		{
			callback(submissions[0]);
		});
	};

	this.updateSubmission = function (id, submission, callback)
	{
		delete submission._id;

		var submissionId = new mongo.ObjectID(id);

		self.db.collection("v1_submission").updateById(submissionId, submission, function (err, submissions)
		{
			callback(submission);
		});
	};
};
