// listly/lib/dao/userDAO.js
var mongo = require("mongodb");
var util = require('util');

var UserDAO = module.exports.UserDAO = function (db)
{
	this.db = db;
};

UserDAO.prototype.findUserById = function (id, callback)
{
	var userId = new mongo.ObjectID(id);

	this.db.collection("v1_user").find({_id:userId}, {limit:1}).toArray(function (err, items)
	{
		callback(items.length > 0 ? items[0] : null);
	});
};

UserDAO.prototype.findUserByUsername = function (username, callback)
{
	this.db.collection("v1_user").find({username:username}, {limit:1}).toArray(function (err, items)
	{
		callback(items.length > 0 ? items[0] : null);
	});
};

UserDAO.prototype.findUserByLogin = function (login, callback)
{
	this.db.collection("v1_user").find({username:login.username, password:login.password}, {limit:1}).toArray(function (err, items)
	{
		callback(items.length > 0 ? items[0] : null);
	});
};

UserDAO.prototype.createUser = function (user, callback)
{
	this.db.collection("v1_user").insert(user, function (err)
	{
		callback(err);
	});
};

UserDAO.prototype.updateUser = function (id, user, callback)
{
	var userId = new mongo.ObjectID(id);

	this.db.collection("v1_user").updateById(userId, user, function (err)
	{
		callback(err);
	});
};
