// listly/lib/routes/loginRoutes.js

var listly = require('listly');
var json = require('json');

var LoginRoutes = module.exports.LoginRoutes = function (listlyServer)
{
	this.userDAO = listlyServer.userDAO;
	this.loginDAO = listlyServer.loginDAO;
	this.listlyEmailService = listlyServer.listlyEmailService;
};

LoginRoutes.prototype.login = function(req,res)
{
	res.render('login/index', listly.renderUtils.params(req,{title: "Login", ngApp: null, noMenu: true}));
};

LoginRoutes.prototype.register = function (req,res)
{
	res.render('register/index', listly.renderUtils.params(req,{title: "Register", ngApp: null, noMenu: true}));
};

LoginRoutes.prototype.doLogin = function(req,res)
{
	var username = req.body.username;
	var password = req.body.password;

	this.loginDAO.loginUser(username,password,function (statusCode,data)
	{
		if (statusCode == 202)
		{
			var user = json.parse(data);
			listly.session.createUserSession(user,req);

			res.writeHead(202);
			res.end();
		}
		else
		{
			res.writeHead(statusCode);
			res.end(data);
		}
	});
};

LoginRoutes.prototype.doRegister = function(req,res)
{
	var username = req.body.username;
	var password = req.body.password;
	var emailAddress = req.body.emailAddress;

	this.loginDAO.registerUser(username,password,emailAddress,function (statusCode,data)
	{
		if (statusCode == 202)
		{
			var user = json.parse(data);
			listly.session.createUserSession(user,req);

			res.writeHead(202);
			res.end();
		}
		else
		{
			res.writeHead(statusCode);
			res.end(data);
		}
	});
};
