// listly/lib/routes/loginRoutes.js

var listly = require('listly');

var LoginRoutes = module.exports.LoginRoutes = function (listlyServer)
{
	this.userDAO = listlyServer.userDAO;
	this.listlyEmailService = listlyServer.listlyEmailService;

	var app = listlyServer.app;
	var rewriter = listlyServer.rewriter;
	app.get('/login', rewriter.rewrite('/views/login'));
	app.get('/register', listly.utils.routeHandler(this,this.registerRedirect));
	app.get('/views/login', listly.utils.routeHandler(this,this.index));
	app.get('/views/register', listly.utils.routeHandler(this,this.registerRedirect));
	app.get('/views/login/login', listly.utils.routeHandler(this,this.login));
	app.get('/views/login/register', listly.utils.routeHandler(this,this.register));
};

LoginRoutes.prototype.registerRedirect = function (req,res)
{
	res.redirect('/views/login#/register');
};

LoginRoutes.prototype.index = function(req,res)
{
	res.render('login/index', listly.renderUtils.params(req,{title: "Login", noMenu: true}));
};

LoginRoutes.prototype.login = function (req,res)
{
	res.render('login/login', listly.renderUtils.params(req,{layout: false}));
};

LoginRoutes.prototype.register = function (req,res)
{
	res.render('login/register', listly.renderUtils.params(req,{layout: false}));
};
