// listly/lib/routes/submissionRoutes.js

var listly = require('listly');

var SubmissionRoutes = module.exports.SubmissionRoutes = function (listlyServer)
{
	this.submissionDAO = listlyServer.submissionDAO;
};

SubmissionRoutes.prototype.home = function(req, res)
{
	res.render('submissions/home/index', listly.renderUtils.params(req,{title: "Latest Submissions"}));
};

SubmissionRoutes.prototype.detail = function (submissionIdParam)
{
	return function (req,res)
	{
		var submissionId = req.params[submissionIdParam];

		res.render('submissions/detail/index', listly.renderUtils.params(req,{title: "Edit Submission", submissionId: submissionId}));
	};
};

SubmissionRoutes.prototype.edit = function (submissionIdParam)
{
	return function (req,res)
	{
		var submissionId = req.params[submissionIdParam];

		res.render('submissions/edit/index', listly.renderUtils.params(req,{title: "Edit Submission", submissionId: submissionId}));
	};
};

SubmissionRoutes.prototype.new = function (req,res)
{
	res.render('submissions/edit/index', listly.renderUtils.params(req,{title: "New Submission", submissionId: null}));
};

SubmissionRoutes.prototype.postSubmission = function (req,res)
{
	var submission = req.body;

	this.submissionDAO.createSubmission(submission, function (err,body)
	{
		res.send(true);
	});
};
