// listly/lib/routes/userRoutes.js

var listly = require('listly');

var UserRoutes = module.exports.UserRoutes = function (listlyServer)
{
	var app = listlyServer.app;
	app.get('/views/user', listly.utils.routeHandler(this,this.index));
	app.get('/views/user/edit', listly.utils.routeHandler(this,this.edit));
	app.get('/views/user/home', listly.utils.routeHandler(this,this.home));
	app.get('/views/user/home/main', listly.utils.routeHandler(this,this.homeMain));
};

UserRoutes.prototype.index = function(req,res)
{
	res.render('user/index', listly.renderUtils.params(req,{title: "Profile"}));
};

UserRoutes.prototype.edit = function(req, res)
{
	res.render('user/edit', listly.renderUtils.params(req,{layout: false}));
};

UserRoutes.prototype.view = function(req, res)
{
	res.render('user/view', listly.renderUtils.params(req,{layout: false}));
};

UserRoutes.prototype.home = function(req, res)
{
	res.render('user/home/index', listly.renderUtils.params(req,{title: "Home"}));
};

UserRoutes.prototype.homeMain = function(req, res)
{
	res.render('user/home/main', listly.renderUtils.params(req,{layout: false}));
};
