// listly/lib/routes/utilRoutes.js

var listly = require('listly');
var json = require('json');

var UtilREST = module.exports.UtilREST = function (listlyServer)
{
	this.listlySocketServer = listlyServer.listlySocketServer;
	this.listlyEmailService = listlyServer.listlyEmailService;
	this.mapReduceService = listlyServer.mapReduceService;

	var app = listlyServer.app;
	app.get('/ping', listly.utils.routeHandler(this,this.doPing));
	app.get('/views/logger', listly.utils.routeHandler(this,this.viewLogger));
	app.get('/fxn/mapReduce', listly.utils.routeHandler(this,this.performMapReduce));
};

UtilREST.prototype.doPing = function (req,res)
{
	this.listlySocketServer.logMsg('ping successful');
	this.listlyEmailService.sendMessage("arieljake@yahoo.com","Ping!","ping successful");

	res.send('ping successful');
};

UtilREST.prototype.viewLogger = function(req,res)
{
	res.render('socket/logger',{layout: false, title: "Logger"});
};

UtilREST.prototype.performMapReduce = function(req,res)
{
	var self = this;

	res.writeHead(200);

	this.mapReduceService.performMapReduce(function (results)
	{
		var output = json.stringify(results);
		res.write(output);
		self.listlySocketServer.logMsg(output);

	}, function ()
	{
		res.end();
	});
};
