// listly/index.js
// server components
module.exports.ListlyServer = require('./lib/server/server').ListlyServer;
module.exports.ListlyRouter = require('./lib/server/router').ListlyRouter;
module.exports.authRequired = require('./lib/server/authRequired');
module.exports.session = require('./lib/server/session');
module.exports.renderUtils = require('./lib/view/renderUtils');
module.exports.settings = require('./lib/server/settings');
module.exports.utils = require('./lib/utils');

// dao
module.exports.ListDAO = require('./lib/dao/listDAO').ListDAO;
module.exports.SubmissionDAO = require('./lib/dao/submissionDAO').SubmissionDAO;
module.exports.UserDAO = require('./lib/dao/userDAO').UserDAO;

// routes/REST
module.exports.ListREST = require('./lib/routes/listREST').ListREST;
module.exports.ListRoutes = require('./lib/routes/listRoutes').ListRoutes;
module.exports.LoginREST = require('./lib/routes/loginREST').LoginREST;
module.exports.LoginRoutes = require('./lib/routes/loginRoutes').LoginRoutes;
module.exports.SubmissionREST = require('./lib/routes/submissionREST').SubmissionREST;
module.exports.SubmissionRoutes = require('./lib/routes/submissionRoutes').SubmissionRoutes;
module.exports.UserREST = require('./lib/routes/userREST').UserREST;
module.exports.UserRoutes = require('./lib/routes/userRoutes').UserRoutes;
module.exports.UtilREST = require('./lib/routes/utilRoutes').UtilREST;

// providers
module.exports.ListlyAmazonSES = require('./lib/server/amazonSES').ListlyAmazonSES;
module.exports.ListlyIronMQ = require('./lib/server/ironMQ').ListlyIronMQ;

// services
module.exports.ListlySocketServer = require('./lib/server/socketServer').ListlySocketServer;
module.exports.ListlyEmailService = require('./lib/server/email').ListlyEmailService;
module.exports.ListlyMessageService = require('./lib/server/messages').ListlyMessageService;
module.exports.MapReduceService = require('./lib/db/mapReduce').MapReduceService;
