// listly/index.js

var router = require('./lib/server/router');
var util = require('util');

module.exports.submissionDAO = require('./lib/dao/submissionDAO');
module.exports.listDAO = require('./lib/dao/listDAO');
module.exports.userDAO = require('./lib/dao/userDAO');

module.exports.mapReduce = require('./lib/db/mapReduce');

module.exports.authRequired = require('./lib/server/authRequired');
module.exports.socketServer = require('./lib/server/socketServer');
module.exports.sessionSettings = require('./lib/server/sessionSettings');

module.exports.renderUtils = require('./lib/view/renderUtils');

module.exports.server = new function ()
{
	var self = this;
	var serverParts = null;

	this.init = function (serverParts)
	{
		self.serverParts = serverParts;
		self.router = router.createRouter(serverParts);
		self.router.init(serverParts);

		// var scheduledTasks = listly.scheduledTasks;
	};

	this.logObj = function (obj)
	{
		var msg = util.inspect(obj);
		self.logMsg(msg);
	};

	this.logMsg = function (msg)
	{
		self.serverParts.listlySocketServer.logMsg(msg);
	};
};
