// listly/index.js
module.exports.utils = require('./lib/utils');

// server components
module.exports.ListlyDBManager = require('./lib/db/dbManager').ListlyDBManager;
module.exports.ListlyServer = require('./lib/server/server').ListlyServer;
module.exports.ListlyAppServer = require('./lib/server/appServer').ListlyAppServer;
module.exports.ListlyAppRouter = require('./lib/server/appRouter').ListlyAppRouter;
module.exports.ListlyDataServer = require('./lib/server/dataServer').ListlyDataServer;
module.exports.ListlyDataRouter = require('./lib/server/dataRouter').ListlyDataRouter;
module.exports.ListlyAuthProvider = require('./lib/server/authProvider').ListlyAuthProvider;
module.exports.authRequired = require('./lib/server/authRequired');
module.exports.session = require('./lib/server/session');
module.exports.renderUtils = require('./lib/view/renderUtils');
module.exports.settings = require('./lib/server/settings');

// dao
module.exports.ListDAO = require('./lib/dao/listDAO').ListDAO;
module.exports.SubmissionDAO = require('./lib/dao/submissionDAO').SubmissionDAO;
module.exports.UserDAO = require('./lib/dao/userDAO').UserDAO;
module.exports.IndexesDAO = require('./lib/dao/indexesDAO').IndexesDAO;

// routes
module.exports.ListRoutes = require('./lib/routes/listRoutes').ListRoutes;
module.exports.LoginRoutes = require('./lib/routes/loginRoutes').LoginRoutes;
module.exports.SubmissionRoutes = require('./lib/routes/submissionRoutes').SubmissionRoutes;
module.exports.UserRoutes = require('./lib/routes/userRoutes').UserRoutes;
module.exports.UtilRoutes = require('./lib/routes/utilRoutes').UtilRoutes;

// REST
module.exports.ListREST = require('./lib/rest/listREST').ListREST;
module.exports.LoginREST = require('./lib/rest/loginREST').LoginREST;
module.exports.SubmissionREST = require('./lib/rest/submissionREST').SubmissionREST;
module.exports.UserREST = require('./lib/rest/userREST').UserREST;

// providers
module.exports.ListlyAmazonSES = require('./lib/server/amazonSES').ListlyAmazonSES;
module.exports.ListlyIronMQ = require('./lib/server/ironMQ').ListlyIronMQ;

// services
module.exports.ListlySocketServer = require('./lib/server/socketServer').ListlySocketServer;
module.exports.ListlyEmailService = require('./lib/server/email').ListlyEmailService;
module.exports.ListlyMessageService = require('./lib/server/messages').ListlyMessageService;
module.exports.MapReduceService = require('./lib/db/mapReduce').MapReduceService;
