// listly/submissionDAO.js
module.exports.findAllSubmissions = function (db,callback)
{
	db.collection("v1_submission").find().toArray(function (err,items) {
		callback(items);
	});
};

module.exports.findSubmissionById = function(db,id,callback)
{
	db.collection("v1_submission").find({_id: id},{limit: 1}).toArray(function (err,items) {
		 callback(items[0]);
	});
};

module.exports.findSubmissionListsById = function(db,id,callback)
{
	db.collection("v1_submission_to_lists").find({_id: id},{limit: 1}).toArray(function (err,items) {
		callback(items[0].value.lists);
	});
};

module.exports.findSubmissionsByTitle = function(db,title,callback)
{
	db.collection("v1_submission").find({title: title}).toArray(function (err,items) {
		callback(items);
	});
};

module.exports.createSubmission = function(db,submission,callback)
{
	db.collection("v1_submission").insert(submission, function (err,submissions)
	{
		callback(submissions[0]);
	});
};

module.exports.updateSubmission = function(db,id,submission,callback)
{
	delete submission._id;

	db.collection("v1_submission").updateById(id, submission, function (err,submissions)
	{
		callback(submission);
	});
};
