// listly/lib/rest/loginREST.js

var listly = require('listly');
var json = require('json');

var LoginREST = module.exports.LoginREST = function (listlyServer)
{
	this.userDAO = listlyServer.userDAO;
	this.listlyEmailService = listlyServer.listlyEmailService;
};

LoginREST.prototype.doLogin = function(req, res)
{
	var login = {
		username: req.body.username,
		password: req.body.password
	};

	this.userDAO.findUserByLogin(login,function(user)
	{
		if (user !== null)
		{
			delete user.password;

			res.writeHead(202);
			res.end(json.stringify(user));
		}
		else
		{
			res.writeHead(207);
			res.end('login failed, please try again');
		}
	});
};

LoginREST.prototype.doRegister = function (req,res)
{
	var self = this;
	var user = req.body;

	this.userDAO.createUser(user, function(user,message)
	{
		if (user !== null)
		{
			listly.session.createUserSession(user,req);

			if (listly.settings.sendRegistrationEmails)
				self.listlyEmailService.sendRegistrationEmail(user);

			res.writeHead(202);
			res.end('registration successful');
		}
		else
		{
			res.writeHead(207);
			res.end(message);
		}
	});
};
