// listly/lib/server/appRouter.js


var listly = require('listly');
var express = require('express');

var ListlyAppRouter = module.exports.ListlyAppRouter = function (listlyServer)
{
	var self = this;

	this.app = listlyServer.app;
	this.rewriter = listlyServer.rewriter;

	this.loginRoutes = new listly.LoginRoutes(listlyServer);
	this.loginREST = new listly.LoginREST(listlyServer);
	this.submissionRoutes = new listly.SubmissionRoutes(listlyServer);
	this.listRoutes = new listly.ListRoutes(listlyServer);
	this.userRoutes = new listly.UserRoutes(listlyServer);
	this.utilRoutes = new listly.UtilRoutes(listlyServer);

	// static
	this.app.get('/favicon.ico', listly.utils.routeHandler(this,this.renderStatic));

	// loginRoutes
	this.app.get('/login', listly.utils.routeHandler(this.loginRoutes,this.loginRoutes.index));
	this.app.get('/register', listly.utils.routeHandler(this.loginRoutes,this.loginRoutes.registerRedirect));

	// listRoutes
	this.app.get('/lists', listly.utils.routeHandler(this.listRoutes,this.listRoutes.main));
	this.app.get('/lists/:listId', listly.utils.routeHandler(this.listRoutes,this.listRoutes.detail));

	// submissionRoutes
	this.app.get('/submissions', listly.utils.routeHandler(this.submissionRoutes,this.submissionRoutes.main));
	this.app.get('/submissions/:submissionId', listly.utils.routeHandler(this.submissionRoutes,this.submissionRoutes.detail));

	// utilRoutes
	this.app.get('/ping', listly.utils.routeHandler(this.utilRoutes,this.utilRoutes.doPing));
	this.app.get('/logger', listly.utils.routeHandler(this.utilRoutes,this.utilRoutes.viewLogger));
	this.app.get('/fxn/mapReduce', listly.utils.routeHandler(this.utilRoutes,this.utilRoutes.performMapReduce));
	this.app.get(/^\/views\//, listly.utils.routeHandler(this,this.renderDefaultViewPartial));

	// userRoutes (last so others can match first)
	this.app.get('/', listly.utils.routeHandler(this.userRoutes,this.userRoutes.home));
	this.app.get('/:username', listly.utils.routeHandler(this.userRoutes,this.userRoutes.profile("username")));
};

ListlyAppRouter.prototype.renderDefaultViewPartial = function (req,res)
{
	res.render(req.url.replace(/^\/views\//,''), listly.renderUtils.params(req,{layout: false}));
};

ListlyAppRouter.prototype.renderStatic = function (req,res,next)
{
	var options = {};
	options.path = req.url;
	options.getOnly = true;

	express.static.send(req,res,next,options);
};
