// listly/lib/server/appServer.js

var listly = require('listly');
var util = require('util');

var ListlyAppServer = module.exports.ListlyAppServer = function (serverParts)
{
	listly.ListlyServer.apply(this, arguments);

	this.socketServer = serverParts.socketServer;

	var remoteDAOSettings = listly.settings.remoteDAOSettings;
	this.userDAO = new listly.UserRemoteDAO(remoteDAOSettings);
	this.submissionDAO = new listly.SubmissionRemoteDAO(remoteDAOSettings);
	this.listDAO = new listly.ListRemoteDAO(remoteDAOSettings);
	this.loginDAO = new listly.LoginRemoteDAO(remoteDAOSettings);

	this.listlyEmailService = new listly.ListlyEmailService(this.emailProvider, this.rootDir + "/resources/emails");
	this.listlySocketServer = new listly.ListlySocketServer(this.socketServer);
	this.emailProvider = new listly.ListlyAmazonSES();

	this.router = new listly.ListlyAppRouter(this);
	// var scheduledTasks = listly.scheduledTasks;
};

listly.utils.inherit(ListlyAppServer, listly.ListlyServer);

ListlyAppServer.prototype.initRoutes = function ()
{
	this.router.initRoutes();
};

ListlyAppServer.prototype.getLoginHandler = function ()
{
	var router = this.router;

	return function (req,res)
	{
		var loginRoutes = router.loginRoutes;
		loginRoutes.doLogin(req,res);
	};
}

ListlyAppServer.prototype.getRegisterHandler = function ()
{
	var router = this.router;

	return function (req,res)
	{
		var loginRoutes = router.loginRoutes;
		loginRoutes.doRegister(req,res);
	};
};

ListlyAppServer.prototype.logObj = function (obj)
{
	var msg = util.inspect(obj);
	this.logMsg(msg);
};

ListlyAppServer.prototype.logMsg = function (msg)
{
	this.listlySocketServer.logMsg(msg);
};
