// listly/lib/server/appServer.js

var listly = require('listly');
var util = require('util');

var ListlyAppServer = module.exports.ListlyAppServer = function (serverParts)
{
	listly.ListlyServer.apply(this, arguments);

	this.socketServer = serverParts.socketServer;

	this.userDAO = new listly.UserRemoteDAO(listly.settings.remoteDAOSettings);
	this.submissionDAO = new listly.SubmissionRemoteDAO(listly.settings.remoteDAOSettings);
	this.listDAO = new listly.ListRemoteDAO(listly.settings.remoteDAOSettings);

	this.listlyEmailService = new listly.ListlyEmailService(this.emailProvider, this.rootDir + "/resources/emails");
	this.listlySocketServer = new listly.ListlySocketServer(this.socketServer);
	this.emailProvider = new listly.ListlyAmazonSES();
	this.router = new listly.ListlyAppRouter(this);

	// var scheduledTasks = listly.scheduledTasks;
};

listly.utils.inherit(ListlyAppServer, listly.ListlyServer);


ListlyAppServer.prototype.logObj = function (obj)
{
	var msg = util.inspect(obj);
	this.logMsg(msg);
};

ListlyAppServer.prototype.logMsg = function (msg)
{
	this.listlySocketServer.logMsg(msg);
};
