// listly/lib/server/dataRouter.js


var listly = require('listly');

var ListlyDataRouter = module.exports.ListlyDataRouter = function (listlyServer)
{
	var self = this;

	this.app = listlyServer.app;
	this.rewriter = listlyServer.rewriter;

	this.submissionREST = new listly.SubmissionREST(listlyServer);
	this.listREST = new listly.ListREST(listlyServer);
	this.userREST = new listly.UserREST(listlyServer);

	// loginREST
	this.app.get('/', function(req, res, next) {
		res.end('home, logged in? ' + !!req.session.user);
	});

	this.app.get('/login', function(req, res, next) {
		if(req.session.user) {
			res.writeHead(303, {Location: '/'});
			return res.end();
		}

		var next_url = req.query.next ? req.query.next : '/';

		res.end('<html><form method="post" action="/login"><input type="hidden" name="next" value="' + next_url + '"><input type="text" placeholder="username" name="username"><input type="password" placeholder="password" name="password"><button type="submit">Login</button></form>');
	});

	this.app.post('/login', function(req, res, next) {
		req.session.user = req.body.username;

		res.writeHead(303, {Location: req.body.next || '/'});
		res.end();
	});

	this.app.get('/logout', function(req, res, next) {
		req.session.destroy(function(err) {
			res.writeHead(303, {Location: '/'});
			res.end();
		});
	});

	this.app.get('/secret', function(req, res, next) {
		if(req.session.user) {
			res.end('proceed to secret lair, extra data: ' + JSON.stringify(req.session.data));
		} else {
			res.writeHead(403);
			res.end('no');
		}
	});

	// listREST
	this.app.get('/lists', listly.utils.routeHandler(this.listREST,this.listREST.getAllLists));
	this.app.get('/lists/:listId', listly.utils.routeHandler(this.listREST,this.listREST.getListById("listId")));

	// submissionREST
	this.app.get('/submissions', listly.utils.routeHandler(this.submissionREST,this.submissionREST.getAllSubmissions));
	this.app.get('/submissions/:submissionId', listly.utils.routeHandler(this.submissionREST,this.submissionREST.getSubmissionById("submissionId")));
	this.app.get('/submissions/:submissionId/lists', listly.utils.routeHandler(this.submissionREST,this.submissionREST.getSubmissionListsById("submissionId")));
	this.app.get('/submissions/search?title=:submissionTitle', listly.utils.routeHandler(this.submissionREST,this.submissionREST.getSubmissionsByTitle("submissionTitle")));
	this.app.post('/submissions', listly.utils.routeHandler(this.submissionREST,this.submissionREST.saveSubmission));

	// userREST
	this.app.get('/users/:userId', listly.utils.routeHandler(this.userREST,this.userREST.getUser('userId')));
	this.app.put('/users/:userId', listly.utils.routeHandler(this.userREST,this.userREST.updateUser('userId')));

};
