// listly/lib/server/dataRouter.js


var listly = require('listly');

var ListlyDataRouter = module.exports.ListlyDataRouter = function (listlyServer)
{
	var self = this;

	this.app = listlyServer.app;
	this.rewriter = listlyServer.rewriter;

	this.submissionREST = new listly.SubmissionREST(listlyServer);
	this.listREST = new listly.ListREST(listlyServer);
	this.userREST = new listly.UserREST(listlyServer);

	/**
	 * REST
 	 */

	// listREST
	this.app.get('/lists', listly.utils.routeHandler(this.listREST,this.listREST.getAllLists));
	this.app.get('/lists/:listId', listly.utils.routeHandler(this.listREST,this.listREST.getListById("listId")));

	// submissionREST
	this.app.get('/submissions', listly.utils.routeHandler(this.submissionREST,this.submissionREST.getAllSubmissions));
	this.app.get('/submissions/:submissionId', listly.utils.routeHandler(this.submissionREST,this.submissionREST.getSubmissionById("submissionId")));
	this.app.get('/submissions/:submissionId/lists', listly.utils.routeHandler(this.submissionREST,this.submissionREST.getSubmissionListsById("submissionId")));
	this.app.get('/submissions/search?title=:submissionTitle', listly.utils.routeHandler(this.submissionREST,this.submissionREST.getSubmissionsByTitle("submissionTitle")));
	this.app.post('/submissions', listly.utils.routeHandler(this.submissionREST,this.submissionREST.saveSubmission));

	// userREST
	this.app.get('/users/:userId', listly.utils.routeHandler(this.userREST,this.userREST.getUser('userId')));
	this.app.put('/users/:userId', listly.utils.routeHandler(this.userREST,this.userREST.updateUser('userId')));

};
