// listly/lib/server/dataServer.js

var util = require('util');
var listly = require('listly');

var ListlyDataServer = module.exports.ListlyDataServer = function (serverParts)
{
	this.app = serverParts.app;
	this.db = serverParts.db;
	this.socketServer = serverParts.socketServer;
	this.rewriter = serverParts.rewriter;
	this.rootDir = serverParts.rootDir;

	this.emailProvider = new listly.ListlyAmazonSES();
	this.messageProvider = new listly.ListlyIronMQ();

	this.listlyEmailService = new listly.ListlyEmailService(this.emailProvider, this.rootDir + "/resources/emails");
	this.listlySocketServer = new listly.ListlySocketServer(this.socketServer);
	this.listlyMessageService = new listly.ListlyMessageService(this.messageProvider);
	this.mapReduceService = new listly.MapReduceService(this.db);

	this.submissionDAO = new listly.SubmissionDAO(this.db);
	this.listDAO = new listly.ListDAO(this.db);
	this.userDAO = new listly.UserDAO(this.db);

	this.router = new listly.ListlyDataRouter(this);
};

ListlyDataServer.prototype.logObj = function (obj)
{
	var msg = util.inspect(obj);
	this.logMsg(msg);
};

ListlyDataServer.prototype.logMsg = function (msg)
{
	this.listlySocketServer.logMsg(msg);
};
