// listly/lib/server/dataRouter.js


var listly = require('listly');

var ListlyDataRouter = module.exports.ListlyDataRouter = function (listlyServer)
{
	var self = this;

	this.app = listlyServer.app;
	this.rewriter = listlyServer.rewriter;

	this.submissionREST = listlyServer.submissionREST;
	this.listREST = listlyServer.listREST;
	this.userREST = listlyServer.userREST;
	this.loginREST = listlyServer.loginREST;

	// listREST
	this.app.get('/lists', listly.utils.routeHandler(this.listREST,this.listREST.getAllLists));
	this.app.get('/lists/:listId', listly.utils.routeHandler(this.listREST,this.listREST.getListById("listId")));

	// submissionREST
	this.app.get('/submissions', listly.utils.routeHandler(this.submissionREST,this.submissionREST.getAllSubmissions));
	this.app.get('/submissions/:submissionId', listly.utils.routeHandler(this.submissionREST,this.submissionREST.getSubmissionById("submissionId")));
	this.app.get('/submissions/:submissionId/lists', listly.utils.routeHandler(this.submissionREST,this.submissionREST.getSubmissionListsById("submissionId")));
	this.app.get('/submissions/search?title=:submissionTitle', listly.utils.routeHandler(this.submissionREST,this.submissionREST.getSubmissionsByTitle("submissionTitle")));
	this.app.post('/submissions', listly.utils.routeHandler(this.submissionREST,this.submissionREST.saveSubmission));

	// userREST
	this.app.get('/users/@:username', listly.utils.routeHandler(this.userREST,this.userREST.getUserByUsername('username')));
	this.app.put('/users/@:username', listly.utils.routeHandler(this.userREST,this.userREST.updateUserByUsername('username')));

	// loginREST
	this.app.post('/login', listly.utils.routeHandler(this.loginREST,this.loginREST.doLogin));
	this.app.post('/register', listly.utils.routeHandler(this.loginREST,this.loginREST.doRegister));

};
