// listly/lib/server/email.js

var amazonSES = require('./amazonSES');
var fs = require('fs');

module.exports.createEmailService = function (rootDir)
{
	return new ListlyEmailService(amazonSES.createEmailProvider(),rootDir);
};

var ListlyEmailService = function (emailProvider,rootDir)
{
	var self = this;

	this.emailProvider = emailProvider;
	this.rootDir = rootDir;
};

ListlyEmailService.prototype.sendMessage = function (to,subject,text)
{
	this.emailProvider.sendMessage({
		from: "ariel@deskmasters.com",
		replyTo: ["ariel@deskmasters.com"],
		to: [to],
		subject: subject,
		text: text,
		html: text
	});
};

ListlyEmailService.prototype.sendRegistrationEmail = function (user)
{
	var self = this;

	fs.readFile(this.rootDir + "/resources/registrationEmail.txt","UTF-8",function (err,data)
	{
		if (err)
		{

		}
		else
		{
			var message = data.toString().replace("#username#",user.username);
			message = message.replace("#userUrl#","http://localhost:3000/views/profile");

			self.sendMessage(
				"arieljake@yahoo.com",
				"Welcome to Listly!",
				message
			);
		}
	});
};
