// listly/lib/server/ironMQ.js
/**
 *
 * /projects/{Project ID}/queues	GET	List Message Queues
 /projects/{Project ID}/queues/{Queue Name}	GET	Get Info About a Message Queue
 /projects/{Project ID}/queues/{Queue Name}/clear	POST	Clear All Messages from a Queue
 /projects/{Project ID}/queues/{Queue Name}/messages	POST	Add a Message to a Queue
 /projects/{Project ID}/queues/{Queue Name}/messages	GET	Get a Message from a Queue
 /projects/{Project ID}/queues/{Queue Name}/messages/{Message ID}	DELETE	Delete a Message from a Queue

 * @type {*}
 */

var IronMQ = require('ironmq');

module.exports.createMessageProvider = function ()
{
	return new ListlyIronMQ();
};

var ListlyIronMQ = function ()
{
	var self = this;

	this.token = "POpDAlGQZi_w6MStfpbYfpjQLp4";
	this.projectId = "4facd5080c389d026800cf6b";
};

ListlyIronMQ.prototype.getConnection = function ()
{
	return IronMQ(this.token).projects(this.projectId);
};

// function callBack(err, obj)
ListlyIronMQ.prototype.putMessage = function (queue,message,callback)
{
	if (callback == undefined)
		callback = function () {};

	this.getConnection().queues(queue).put(message,callback);
};

// function callBack(err, msgs)
ListlyIronMQ.prototype.getMessage = function (queue,callback)
{
	if (callback == undefined)
		callback = function () {};

	this.getConnection().queues(queue).get(callback);
};

// function callBack(err, obj)
ListlyIronMQ.prototype.delMessage = function (queue,messageId,callback)
{
	if (callback == undefined)
		callback = function () {};

	this.getConnection().queues(queue).del(messageId,callback);
};
