// listly/lib/server/session.js

var connectMongo = require('connect-mongo');
var underscore = require('underscore');

module.exports.createSessionSettings = function (express)
{
	var mongoStore = connectMongo(express);

	return {
		secret: "listlyisfun",
		store: new mongoStore({
								  db: "listly",
								  collection: "v1_session",
								  host: "127.0.0.1",
								  port: 27017,
								  auto_reconnect: true,
								  clear_interval: 300,
								  stringify: true
							  })
	};
};

module.exports.isSessionActive = function (req)
{
	return req.session.user !== undefined;
}

module.exports.createUserSession = function (user,req)
{
	var sessionUser = underscore.clone(user);
	delete sessionUser.password;

	req.session.user = sessionUser;
};

module.exports.clearUserSession = function (req)
{
	req.session.user = undefined;
};

module.exports.getCurrentUsername = function (req)
{
	if (req.session !== undefined && req.session.user !== undefined)
		return req.session.user.username;
	else
		return undefined;
}
