// listly/lib/daoRemote/remoteDAO.js
var needle = require('needle');
var underscore = require('underscore');

var RemoteDAO = module.exports.RemoteDAO = function (params)
{
	if (params)
	{
		this.defaultOptions = params.defaultOptions;
		this.baseUrl = params.baseUrl;
	}
};

RemoteDAO.prototype.prepareUrl = function (path)
{
	return this.baseUrl + path;
};

RemoteDAO.prototype.prepareOptions = function (options)
{
	if (options == null)
		options = {};

	underscore.defaults(options, this.defaultOptions);

	return options;
};

RemoteDAO.prototype.post = function (path, data, opts, callback)
{
	var url = this.prepareUrl(path);
	var options = this.prepareOptions(opts);

	needle.post(url, data, options, callback);
};

RemoteDAO.prototype.get = function (path, opts, callback)
{
	var url = this.prepareUrl(path);
	var options = this.prepareOptions(opts);

	needle.get(url, options, callback);
};

RemoteDAO.prototype.put = function (path, data, opts, callback)
{
	var url = this.prepareUrl(path);
	var options = this.prepareOptions(opts);

	needle.put(url, data, options, callback);
};

RemoteDAO.prototype.head = function (path, opts, callback)
{
	var url = this.prepareUrl(path);
	var options = this.prepareOptions(opts);

	needle.head(url, options, callback);
};

RemoteDAO.prototype.delete = function (path, opts, callback)
{
	var url = this.prepareUrl(path);
	var options = this.prepareOptions(opts);

	needle.delete(url, options, callback);
};
